torch/distributed/_tensor/placement_types.py

```diff
@@ -378,13 +378,20 @@ class DTensorSpec:
 
     def __hash__(self) -> int:
         # hashing and equality check for DTensorSpec are used to cache the sharding
-        # propagation results. We only need to consider the mesh, placements and shape
+        # propagation results. We only need to consider the mesh, placements, shape
+        # dtype and stride.
         # Caveat: we need to keep this in mind and sync hash and eq if we add more
-        # fields to them,
+        # fields to them.
         if self.tensor_meta is not None:
-            return hash((self.mesh, self.placements, self.tensor_meta.shape))
-        else:
-            return hash((self.mesh, self.placements))
+            return hash(
+                (
+                    self.mesh,
+                    self.placements,
+                    self.tensor_meta.shape,
+                    self.tensor_meta.dtype,
```

Comment:
why we need to hash the dtype and stride?

Also if self.tensor_meta is None then we don't hash, although every dtensor spec should be populated with this field, it would be better if we raise a error if this becomes None

Reply:
Oh, because for the sharding prop, if we don't include stride, the transpose op will get wrong results, so that in a simple TP MLP example, we run into errors in the backward. That's why originally, we didn't figure out this in the non-periodic unit tests.

Reply:
> Also if self.tensor_meta is None then we don't hash

yes. let me update the PR with all local changes.

Reply:
I see, did we try to update both __hash__ and __eq__ and only compare with shape (i.e. deleting dtype/stride in __eq__ too) and whether that works?

Reply:
No, that will cause errors. Because, we need stride to differentiate between different op_schema, specifically, the DTensorSpec within args_schema.

Reply:
hmm I see, do you have some example about DTensorSpecs that have the same placement, shape, but different stride,  and they need different output sharding? I am curious to see why that would cause errors.